controllers/nim_account_controller.go

```diff
@@ -0,0 +1,428 @@
+	if err != nil {
+		return nil, err
+	}
+
+	return cmap, nil
+}
+
+// reconcileRuntimeTemplate is used for reconciling the template encapsulating the serving runtime
+func (r *NimAccountReconciler) reconcileRuntimeTemplate(ctx context.Context, account *v1.Account) (*templatev1.Template, error) {
+	template := &templatev1.Template{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      constants.NimRuntimeTemplateName,
+			Namespace: account.Namespace,
+		},
+	}
+
+	if _, err := controllerutil.CreateOrUpdate(ctx, r.Client, template, func() error {
+		if err := controllerutil.SetControllerReference(account, template, r.Scheme()); err != nil {
+			return err
+		}
+
+		template.Annotations = map[string]string{
+			"opendatahub.io/apiProtocol":         "REST",
+			"opendatahub.io/modelServingSupport": "[\"single\"]",
+		}
+
+		template.Objects = []runtime.RawExtension{
+			{Object: utils.GetNimServingRuntimeTemplate()},
+		}
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	return template, nil
+}
+
+// reconcileNimPullSecret is used to reconcile the pull secret for pulling the custom runtime images
+func (r *NimAccountReconciler) reconcileNimPullSecret(
+	ctx context.Context, ownerCfg *ssametav1.OwnerReferenceApplyConfiguration, namespace, apiKey string,
+) (*corev1.Secret, error) {
+	creds := map[string]map[string]map[string]string{
+		"auths": {
+			"nvcr.io": {
+				"username": "$oauthtoken",
+				"password": apiKey,
+			},
+		},
+	}
+
+	credsJson, marshErr := json.Marshal(creds)
+	if marshErr != nil {
+		return nil, marshErr
+	}
+
+	secretCfg := ssacorev1.
+		Secret(constants.NimPullSecretName, namespace).
+		WithData(map[string][]byte{".dockercfg": credsJson}).
+		WithType(corev1.SecretTypeDockercfg).
+		WithOwnerReferences(ownerCfg)
+
+	secret, err := r.KClient.CoreV1().Secrets(namespace).
+		Apply(ctx, secretCfg, metav1.ApplyOptions{FieldManager: constants.NimApplyConfigFieldManager, Force: true})
+	if err != nil {
+		return nil, err
+	}
+	return secret, nil
+}
+
+// updateStatus is used for fetching an updating the status of the account
+func (r *NimAccountReconciler) updateStatus(ctx context.Context, subject types.NamespacedName, status v1.AccountStatus) {
+	logger := log.FromContext(ctx)
+	logger.V(1).Info("updating status")
+
+	account := &v1.Account{}
+	if err := r.Client.Get(ctx, subject, account); err != nil {
+		if !k8serrors.IsNotFound(err) {
+			logger.Error(err, "failed to fetch account for status update")
+		}
+	} else {
+		account.Status = *status.DeepCopy()
+		if err = r.Client.Status().Update(ctx, account); err != nil {
+			logger.Error(err, "failed to update account status")
+		}
+	}
+}
+
+// createOwnerReferenceCfg is used to create an owner reference config to use with server side apply
+func (r *NimAccountReconciler) createOwnerReferenceCfg(account *v1.Account) *ssametav1.OwnerReferenceApplyConfiguration {
+	gvks, _, _ := r.Scheme().ObjectKinds(account)
```

Comment:
nit: given the usage path, at this point probably the GVK of the account should be set so it is probably not needed to resolve it via the scheme.

Reply:
note: this is not a blocker, just a suggestion but it seems there are some issues that prevent to implement it as described

Reply:
Added a comment referring this in https://github.com/opendatahub-io/odh-model-controller/pull/289/commits/53cf8faa9f79792e0ab12174d53bc67bccc3b3b7.

Basically, what happens is that while working as expected in a live cluster, when running using envtest, the object's TypeMeta is being stripped. See the attached screenshots:

**Live**
![image](https://github.com/user-attachments/assets/147ea2f4-0fa8-4b72-88b1-4c80519a41a9)

**EnvTest*
![image](https://github.com/user-attachments/assets/e49c8d7d-1da2-45ff-b244-f880eaad5513)

Reply:
This was tested with the latest envtest (0.19.1) using kube 1.26, 1.29, and 1.30.